### ISSUE TITLE
Deterministic behavior for phasespace generator

### ISSUE THREAD

**Issue by spflueger** [ComPWA/tensorwaves#59]
The phase space generator using the `phasespace` package has to be improved to accept a random generator. This will help to ensure deterministic behavior based on the seed. Or at least the deterministic behavior has to be verified.

So TFPhaseSpaceGenerator and TFUniformRealNumberGenerator are currently not decoupled, because the `phasespace` package uses `tf.random` internally as well. This is currently in conflict with our generate definition and can be problematic. The random generators have to be kept in sync. We might have to think about our generate definition as well.

**[Estimated effort: 1-5 days]**

**Comment by redeboer** [comment]
The `phasespace` package makes uses of [`tf.random.uniform`](https://www.tensorflow.org/api_docs/python/tf/random/uniform)
https://github.com/zfit/phasespace/blob/6e442c7644482120c0efe42f0a85ed46890da364/phasespace/phasespace.py#L326

This is an old RNG that TensorFlow  [discourages from TF1.x on](https://www.tensorflow.org/guide/random_numbers). Instead they recommend using [`tf.random.Generator`](https://www.tensorflow.org/api_docs/python/tf/random/Generator).

**Comment by mayou36** [comment]
This is indeed outdated and should be updated. I am not quite sure myself for the best strategy of this however.

I think the most pragmatic solution is to have a global generator that can be set (for any package). This would fulfill the two requirements we have (usually):
- random numbers are random and not correlated (just recently ran into this when doing convolution by addition...). Since one global generator is used, this is fulfilled.
- deterministic behavior: same code with same generator will result in the same result. However, a single model won't output the same with the same seed (since it's a global seed), but then this is usually not required.

Why I would not have a "per instance" generator or something: I don't see a real usecase for this and it just adds to the complexity.

**Comment by spflueger** [comment]
Yes, we need a single source of truth for the random numbers to guarantee the two points you mentioned.

The two possible options are:

1. Global generator as mentioned above
2. Feeding a single random number generator instance to the event generation facade code. The facade code will ensure that only that one rng instance is used.

I prefer option 2, as it is more flexible and modular. Furthermore I believe option 2 is actually also easier to implement, since all that has to be done in the facade is hand the rng instance to the phase space event generator. The only tricky part might be tensorflow itself, that injecting a regular python callable (interface to rng) requires adjustments. But it would be fine if the phasespace package would just allow the tensorflow random generator as an option.

What do you think @redeboer , @mayou36 ?

**Comment by mayou36** [comment]
While I agree that 2. is more modular, it has two drawbacks in my view:
- there is no actual usecase I assume. Why should it be modular, does someone at some point want to inject a random generator? Flexibility is nice, but there should be a foreseeable use-case; in the end we don't build a general statistics library.
- if, then I would rather go to something like TensorFlow Probability does and make the user provide a seed (https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Distribution?version=nightly#sample): this is not only more practical from a TF point-of-view (passing numbers vs passing python object) but is also backend agnostic.

So I would go for 1. with the option to explicitly set the random number generator (at the beginning) and otherwise lazy get one and set it in `phasespace`.

Or what exactly would be the use-case that is not covered with this? Maybe I am missing out some cases.

**Comment by redeboer** [comment]
Some quick thoughts from my side:

* The original motivation for this issue is that we need to be able to generate a deterministic phase space sample (and in the extend, a deterministic uniform distribution) to have reproducible analysis results. I currently don't see how to specify the generator seed with the `phasespace` interfaces.

* I also don't see a use case for inserting a custom generator either, but the reasoning we follow is just formal: keep the responsibilities of each component simple. In addition, keeping a framework flexible and modular (that is, not pinning it down to specific use cases) is always safer in the long term, because you can never 100% anticipate the user's needs. Therefore I would say 2. makes our (the developer's) lives easier.

* Just experimented a bit with `tf.random.Generator`. It seems to me that TF also prefers a more modular approach, which is why they discourage this global seed implementation (see e.g. [`tf.random.set_seed`](https://www.tensorflow.org/api_docs/python/tf/random/set_seed)). This is another reason why I'm not in favour of 1.

For now, I would just replace the call to `tf.random.uniform` in `phasespace` with `tf.random.Generator`. For the design discussion, perhaps it's better to create an issue in the `phasespace` repo.

**Comment by mayou36** [comment]
I agree in general, just to clarify my idea:
- for phasespace a seed for the sampling (like `TFP` does) seems to make sense
- for tensorwave/zfit, I would go for a global one. Because:

> n addition, keeping a framework flexible and modular (that is, not pinning it down to specific use cases) is always safer in the long term, because you can never 100% anticipate the user's needs.

This should not serve as the only argument to whether we make it flexible or not, otherwise one could argue to simply provide Python to the user as it gives "the largest flexibility". TensorFlow, TFP and PhaseSpace are in a way lower level libraries than tensorwave, so some flexibility that the frameworks provide may not need to be propagated to tensorwaves. There is a level where a library should specialize: it means to _reduce_ the flexibility by making assumptions; this is in the end what makes a library useable.

Furhtermore, this sometimes _look_ flexible, but it is now the opposite: e.g. switching the backend is now impossible: if a user gives a TF generator, his script will break in the future if we don't use TF anymore, with a seed this won't happen. I would favor a seed over a generator, it is simply more user friendly and I cannot think of a single usecase where a user wants to give a _generator_; basically it does the same as now, just creating a generator from the seed.

Another problem that we may created now is the static seed `123456`: do we really want to have a static seed as the default? Not a random one _if nothing is set_ (e.g. use `from_non_deterministic_state` instead of a fixed seed)?
In fact, I think this leads now to a large problem: if we sample _multiple times_ with the `generate_data` function, they will be done with the same seed from a new Generator instance, meaning they will be the same. The default expected behavior from any other library - that I am aware of - is to use non-deterministic random numbers (numpy, tensorflow,...) if no seed is given.

Opened a [phasespace issue](https://github.com/zfit/phasespace/issues/36) to implement it downstreams

**Comment by spflueger** [comment]
> I agree in general, just to clarify my idea:
> 
> * for phasespace a seed for the sampling (like `TFP` does) seems to make sense
> * for tensorwave/zfit, I would go for a global one. Because:
> 
> > n addition, keeping a framework flexible and modular (that is, not pinning it down to specific use cases) is always safer in the long term, because you can never 100% anticipate the user's needs.
> 
> This should not serve as the only argument to whether we make it flexible or not, otherwise one could argue to simply provide Python to the user as it gives "the largest flexibility". TensorFlow, TFP and PhaseSpace are in a way lower level libraries than tensorwave, so some flexibility that the frameworks provide may not need to be propagated to tensorwaves. There is a level where a library should specialize: it means to _reduce_ the flexibility by making assumptions; this is in the end what makes a library useable.
> 

But what do we lose by using option 2? As far as I can see we only gain things: being able to switch the rng backend and avoiding the global rng.
I'm not sure its also clear that the facade functions in tensorwaves generate uniform phase space samples (event weights = 1), by doing hit and miss MC. This requires a rng on the level of the facade, as well as in the phase space event generator (which usually generate events with weights). To make this deterministic, using an instance is much more convenient.

> Furhtermore, this sometimes _look_ flexible, but it is now the opposite: e.g. switching the backend is now impossible: if a user gives a TF generator, his script will break in the future if we don't use TF anymore, with a seed this won't happen. I would favor a seed over a generator, it is simply more user friendly and I cannot think of a single usecase where a user wants to give a _generator_; basically it does the same as now, just creating a generator from the seed.
> 
The idea is to define an interface for the rng. The code using rng's would simply accept that interface (as an argument) and things work perfectly fine. Or am I missing something?

> Another problem that we may created now is the static seed `123456`: do we really want to have a static seed as the default? Not a random one _if nothing is set_ (e.g. use `from_non_deterministic_state` instead of a fixed seed)?
> In fact, I think this leads now to a large problem: if we sample _multiple times_ with the `generate_data` function, they will be done with the same seed from a new Generator instance, meaning they will be _the exact same_. The default expected behavior from any other library - that I am aware of - is to use non-deterministic random numbers (numpy, tensorflow,...) if no seed is given.

That is a good point! So we should change the default behavior to a random seed. A bit separate issue though. 
> 
> Opened a [phasespace issue](https://github.com/zfit/phasespace/issues/36) to implement it downstreams

**Comment by mayou36** [comment]
> But what do we lose by using option 2? 
Not a lot, I agree, we just add more arguments to a function, but tbh if it is not used, it can still be removed otherwise.

> As far as I can see we only gain things: being able to switch the rng backend and avoiding the global rng (which is is a nightmare for debugging as there are no boundaries for it).
> I'm not sure its also clear that the facade functions in tensorwaves generate uniform phase space samples (event weights = 1), by doing hit and miss MC. This requires a rng on the level of the facade, as well as in the phase space event generator (which usually generate events with weights). To make this deterministic, using an instance is much more convenient.
> 
Hm, I don't think that really matters but I agree to the flexibility from another point of view In fact, in zfit the random sampler is even more flexible to allow for importance samping etc for the accept reject.

> 
> That is a good point! So we should change the default behavior to a random seed. A bit separate issue though.

Yes, agree, it's separate.

**Comment by spflueger** [comment]
> But it would be fine if the phasespace package would just allow the tensorflow random generator as an option.
Just to clarify what I meant with this sentence. So I'm fine if the random generator passed into `phasespace.generate()` would be limited to some uniform random generator of tensorflow to avoid having to fight with any tf.Tensor related issues.

I guess the uniform phase space generation (weights =1) is a good use case for having the rng separated. Since a new phase space gen would require a new managing facade code to handle the hit and miss for that generator (if the rng cannot be fed into the phase space generator).

Basically I just would like to see some concrete example why passing the rng into the phase space generator is making things worse. I can only see that its not what ppl are used to. In the sense that its expected the phase space generator should be taking care of the randomness factor. However the hit and miss sampling is quite difficult to tackle when deterministic behavior is required.

**Comment by spflueger** [comment]
So I just thought about this a bit more. I think if we just used a separate RNG in tensorwaves to do the hit and miss everything works and that keeps phasespace simple. So everyone should be happy :)

The only requirement is that phasespace needs some way to set a seed to be able to create a deterministic sample.

**Comment by mayou36** [comment]
Yep, I see, however I guess it's worth to take a step back since the whole problematic is by far non-trivial and we have a few libraries (tensorwave, zfit, phasespace, (ampliTF)) that should get it right, so a well thought reasoning would be great. Also to disentangle the views of technicality vs user.

(What I think are) Requirements from a user/dev point of view:
- U1: reproducibility of a _script_ (which contains multiple function calls): a single, human readable seed should be set by the user and the output should be the same. I think the use-case is clear
- U2: If _no_ seed it set, the random number generation should vary each time (as any other library does). use-case: it's the default behavior everywhere else, fixing risks of biasing
- U3: deterministic behavior of a _function_. Given the same seed, the function should return the same values. Use-case: (I assume) debugging, which requires a _human readable_ seed or a _reproducible_ seed.

For the technical requirements:
- T1: unbias ofc everything, random is random etc.
- T2: Prevent bottlenecks from global synchronized writes
- T3: place generator on _same_ device as computation (otherwise copy bottleneck, e.g. CPU-GPU)

Problems with the current implementation (as I see it, maybe I miss code/something):
- U1 is not satisfied, a user needs to take care of _all_ calls to a rng-using function
- U2 (actually not, but we can fix that)
- U3 it is deterministic _if_ we give the same generator again. However, we can't (correct?), because it's not simply a seed number that we can remember. If we have several function calls, we don't know how far the rng advanced, so while for the first call (assuming we create a rng from a seed), this is reproducible, later calls using the same rng are not. Especially since the rng is now directly used inside, it can advance an unkown time.

- (T1: again, when `123456` is removed, better, depends on solution)
- T3: we use an outside created generator that is maybe placed on a different device

My proposal is to use:
- from the user side; seeds (or generators) are allowed: global and local in order to satisfy U1 and U3.
- if a seed is given, a generator is created from it with the seed (and salted by the function name?)
- if nothing is given, a generator is retrieved from the global generator with the `split` that places the generator local and makes advances independent.
(`g = tf.random.get_global_generator().split(1)[0] `)
- if a generator is given, a _new rng_ is retrieved with the split, which prevents the outside rng to advance and be set in the right place.



This leads to:
- U1: use global seed. creates and sets global rng, local rngs from global rng -> deterministic
- U2: fulfilled as global rng is initialized randomly -> everything else random
- U3: fulfilled as a seed can be provided

- T1: works, due to using the `split`, a new rng is used in every function, salting prevents exact same seed reuse
- T2: works as we only retrieve rngs from it
- T3: works as we always use the split (or instantiate in-place from a seed) to get a new rng

What to change: create a rng factory that does that and returns an rnd. And a global `set_seed` (as well as `set_rng`?)

Which libraries: I think `phasespace` should surely implement this behavior. For zfit (and tensorwave), U3 may be overkill/impossible, as there is a lot of randomness involved everywhere and _usually_ a single function call may not be so relevant. However, for the `sampling method`, I can imagine to add that (but e.g. for integration etc not needed).

I know it's a long read, but IMHO worth to have it once-and-for-all discussed and settled. Rather do it right than regret later, we all know the cost of bugs and maintenance ;)

What I found useful sources: [TF explanation of RNGs](https://www.tensorflow.org/guide/random_numbers), the TFP things (`SeedStream` seems a little outdated) and the [jax reasoning of their RNGs](https://github.com/google/jax/blob/master/design_notes/prng.md)

**Comment by redeboer** [comment]
Thanks for your clear exposition and reasoning @mayou36! I think this also answers some of the reservations we have towards using globals: whereas globals are usually a thing to be avoided, the combination of working with multiple devices and having to ensure consistent, deterministic output (if desired) makes it all rather tricky, and working with some global states could be a useful tool here, as is clear from the TF guides.

I'm quite new to the complexities of RNGs, so it's interesting to read up on this, both the technicalities and how `jax` and TF try to deal with this through their interfaces. For instance, it's good to be aware of this splitting of (global) RNGs. I mention it, because it seems to be an important concept in both [`jax.random`](https://jax.readthedocs.io/en/latest/jax.random.html) as in the TF guides, meaning that our interfaces would have to mimic this if they are to be wrappers around such backends.

Probably it's easiest to first tackle this issue in `phasespace` (https://github.com/zfit/phasespace/issues/36, as well as allowing for deterministic behaviour whether through a seed or custom generator)? In `phasespace`, the use-case is most well-defined and its interfaces most stable, so it's clearest what to do there. (TensorWaves is still rather experimental and the design is mostly taken over from ComPWA, which is also why there may be some mistaken C++ reasoning involved here and there.)

@mayou36 If you wish, I can try implementing your proposal in `phasespace` later this week. I'll probably get a few things wrong there, as I'm not as proficient in TF and it could be that I will misunderstand some of the intentions of the rest of the framework, but we can sort that out in the resulting PR. For the issue here in `tensorwaves`, @spflueger and I will get back to it next week.

**Comment by mayou36** [comment]
> ...whereas globals are usually a thing to be avoided...

I absolutely agree of course as a general statement

> Probably it's easiest to first tackle this issue in `phasespace`

Fully agree! If you agree with the strategy/arguments above (so far), then we can have a consistent way of the behavior across libraries and we can put this into `phasespace`, then `tensorwave` can build on this and in `zfit` we may put it like this as well.

I guess I can tackle that for `phasespace` until the end of the week and ping you on the PR, since I got quite an idea

**Comment by spflueger** [comment]
Thanks @mayou36 for the detailed proposal!

I think you cover all important points. I did not know about this splitting feature, which solves the "tricky" part of having independent rngs! Another thing I did not think of is **T3**. I still would like to have the code as DRY as possible, but yet the possibility to chose the computation backend. So I guess the AmpliTF toolkit/framework would have to provide some random definition to allow this. Right?

> My proposal is to use:
> 
> * from the user side; seeds (or generators) are allowed: global and local in order to satisfy U1 and U3.
> * if a seed is given, a generator is created from it with the seed (and salted by the function name?)
> * if nothing is given, a generator is retrieved from the global generator with the `split` that places the generator local and makes advances independent.
>   (`g = tf.random.get_global_generator().split(1)[0] `)
> * if a generator is given, a _new rng_ is retrieved with the split, which prevents the outside rng to advance and be set in the right place.
> 
> This leads to:
> 
> * U1: use global seed. creates and sets global rng, local rngs from global rng -> deterministic
> * U2: fulfilled as global rng is initialized randomly -> everything else random
> * U3: fulfilled as a seed can be provided
> * T1: works, due to using the `split`, a new rng is used in every function, salting prevents exact same seed reuse
> * T2: works as we only retrieve rngs from it
> * T3: works as we always use the split (or instantiate in-place from a seed) to get a new rng
> 
> What to change: create a rng factory that does that and returns an rnd. And a global `set_seed` (as well as `set_rng`?)
> 

Sound good to me! I just need some clarification on one thing. If I understood things correctly, the combination of **U1** and **T1** requires some global rng class that can spawn independent rngs throughout the entire script and at the same time guarantee that all random number business is deterministic.
I wonder if there is some way to get rid of this global rng. I don't have any other solution though...

An implementation detail that is important to me though, is that all of the modules below the user interface require seeds or rng as arguments. So in other words, the modules should NOT have any dependence to the global rng by obtaining some rng or state from the global rng class. Some facade/UI code should handle that the global rng, passes seeds/rngs to the subcomponents.

> Which libraries: I think `phasespace` should surely implement this behavior. For zfit (and tensorwave), U3 may be overkill/impossible, as there is a lot of randomness involved everywhere and _usually_ a single function call may not be so relevant. However, for the `sampling method`, I can imagine to add that (but e.g. for integration etc not needed).
> 

The global rng has to be managed in the most highest level package (tensorwaves/zfit). So phasespace just needs a change that allows to put the tf random in a deterministic state (via a seed or a rng). I think we don't have to worry about having an independent rng at this point. That has to be managed by tensorwaves. Or am I missing something?

**Comment by mayou36** [comment]
Sorry, I missed to answer, but I completely agree on the points and implemented something along the line.

Unfortunately, since the RNG cannot be split in graph mode, the interface now allows for either passing through a RNG instance or a seed. The RNG will be directly used. (This can actually be changed inside by using a from seed, but it doesn't matter at this point, the user interface will remain the same)

The package is published under 1.2.0 on pypi for the changes see here (https://github.com/zfit/phasespace/releases/tag/1.2.0)

**Comment by redeboer** [comment]
Thanks a lot @mayou36, also for your valuable ideas in this issue!
We'll implement `phasespace` 1.2.0 into `tensorwaves` later and let you know if there are issues
